docker/docker_image_dest.go

```diff
@@ -168,12 +174,128 @@ func (d *dockerImageDestination) PutManifest(m []byte) error {
 	}
+
+	for i, signature := range signatures {
+		url := signatureStorageURL(d.c.signatureBase, d.manifestDigest, i)
+		if url == nil {
+			return fmt.Errorf("Internal error: signatureStorageURL with non-nil base returned nil")
+		}
+		err := d.putOneSignature(url, signature)
+		if err != nil {
+			return err
+		}
+	}
+	// Remove any other signatures, if present.
+	// We stop at the first missing signature; if a previous deleting loop aborted
+	// prematurely, this may not clean up all of them, but one missing signature
+	// is enough for dockerImageSource to stop looking for other signatures, so that
+	// is sufficient.
+	for i := len(signatures); ; i++ {
+		url := signatureStorageURL(d.c.signatureBase, d.manifestDigest, i)
+		if url == nil {
+			return fmt.Errorf("Internal error: signatureStorageURL with non-nil base returned nil")
+		}
+		missing, err := d.c.deleteOneSignature(url)
+		if err != nil {
+			return err
+		}
+		if missing {
+			break
+		}
+	}
+
 	return nil
 }
 
+// putOneSignature stores one signature to url.
+func (d *dockerImageDestination) putOneSignature(url *url.URL, signature []byte) error {
+	switch url.Scheme {
+	case "file":
+		logrus.Debugf("Writing to %s", url.Path)
+		err := os.MkdirAll(filepath.Dir(url.Path), 0755)
+		if err != nil {
+			return err
+		}
+		err = ioutil.WriteFile(url.Path, signature, 0644)
+		if err != nil {
+			return err
+		}
+		return nil
+
+	case "http", "https":
+		// FIXME? This will very likely need some kind of authentication support.
+		// Punt and have users configure a file:// path to the webroot?
+		// Also see deleteOneSignature below.
+		req, err := http.NewRequest("PUT", url.String(), bytes.NewReader(signature))
+		if err != nil {
+			return err
+		}
+		logrus.Debugf("%s %s", req.Method, req.URL)
+		res, err := d.c.client.Do(req)
+		if err != nil {
+			return err
+		}
+		defer res.Body.Close()
+		if res.StatusCode != http.StatusOK {
+			return fmt.Errorf("Error writing signature to %s: status %d", url.String(), res.StatusCode)
+		}
+		return nil
+	default:
+		return fmt.Errorf("Unsupported scheme when writing signature to %s", url.String())
+	}
+}
+
+// deleteOneSignature deletes a signature from url, if it exists.
+// If it successfully determines that the signature does not exist, returns (true, nil)
+func (c *dockerClient) deleteOneSignature(url *url.URL) (missing bool, err error) {
+	switch url.Scheme {
+	case "file":
+		logrus.Debugf("Deleting %s", url.Path)
+		err := os.Remove(url.Path)
+		if err != nil && os.IsNotExist(err) {
+			return true, nil
+		}
+		return false, err
+
+	case "http", "https":
+		// See putOnesignature above regarding autentication.
+		req, err := http.NewRequest("DELETE", url.String(), nil)
+		if err != nil {
+			return false, err
+		}
+		logrus.Debugf("%s %s", req.Method, req.URL)
+		res, err := c.client.Do(req)
+		if err != nil {
+			return false, err
+		}
+		defer res.Body.Close()
+		if res.StatusCode == http.StatusNotFound {
```

Comment:
question, is there an API to follow when implementing a remove sigstore? sorry if I missed it but I was just wondering this